### PR TITLE
Custom routes

### PR DESCRIPTION
**Pull request by MoOx**
No more "random" generated/consumed URL.
md-collection-loader now add fields in addition to __url: __resourceUrl (actual file, so might contains /index/html) and __dataUrl (json file url).

.md files accept an "route" field for custom urls.

Consequence: PageContainer now know before fetching if a url is know or unknow.

### CODE REVIEW

**Comment by thangngoc89** [comment]
Bravo :clap:

File: src/builder/index.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ import debug from "debug"
 import webpack from "./webpack"
 import devServer from "./server"
 
-import filenameToUrl from "../filename-to-url"
+import collection from "../md-collection-loader/cache"
```
Comment (thangngoc89) [inline src/builder/index.js]: This is in memory cache. Will it cause invalid checksum in dev server?

Comment (MoOx) [inline src/builder/index.js]: Should not :) And if it create some, we will fix it somehow :)

Comment (thangngoc89) [inline src/builder/index.js]: Ha. It's not

File: src/builder/server.js [hunk 2]
```diff
@@ -119,29 +123,22 @@ export default (webpackConfig, options = {}) => {
...
+        !req.originalUrl.endsWith("/")
+      ) {
         res.redirect(req.originalUrl + "/")
       }
 
+      const relativeUri = item.__dataUrl.replace(config.baseUrl.pathname, "")
+      const filepath = join(config.cwd, config.destination, relativeUri)
+      const fileContent = memoryFs.readFileSync(filepath)
       log(
         `Using '${ filepath }' to pre-render '${ req.originalUrl }' (${ uri })`
```
Comment (thangngoc89) [inline src/builder/server.js]: I think this log line is useless now.

File: src/builder/index.js (hunk 1 above)
Comment (MoOx) [inline src/builder/index.js]: ;)

File: src/builder/index.js [hunk 3]
```diff
@@ -33,24 +33,16 @@ export default function(options) {
...
 
-      // There is probably a better way to get markdown as json without reading
-      // fs, but I am tired
-      const pagesData = {}
+      // TODO use a more reliable way to only get entry points
       const assetsFiles = {
         css: [],
         js: [],
       }
       const assets = stats.compilation.assets
```
Comment (thangngoc89) [inline src/builder/index.js]: Is `stats.compilation.namedChunks` better ?

File: src/builder/index.js [hunk 4]
```diff
@@ -33,24 +33,16 @@ export default function(options) {
...
-        if (name.endsWith("index.json")) {
-          const url = filenameToUrl(name)
-          pagesData[url] = JSON.parse(assets[name]._value)
-        }
-
         if (name.endsWith(".js")) {
           assetsFiles.js.push(name)
         }
-
         if (name.endsWith(".css")) {
```
Comment (thangngoc89) [inline src/builder/index.js]: Also I think this can be `else if` no need for another condition (my fault)

Comment (MoOx) [inline src/builder/index.js]: yeah, no big deal :)

**Comment by MoOx** [comment]
I will fix windows (I used path.join() for url, which is incorrect, so I will create a small util to handle that since there is nothing native in node).

**Comment by thangngoc89** [comment]
`path.sep()` ? It will return `\` in windows and `/` in linux

**Comment by MoOx** [comment]
That what is using path.join(), but url are the same on windows and linux haha.
So I will push something like


```js
const urlJoin = () => (path.join(...args).replace(path.sep, "/"))
```

What do you think?

**Comment by thangngoc89** [comment]
Should replace it everywhere

`const re = new RegExp(path.sep, g)`

**Comment by thangngoc89** [comment]
If I'm offline and navigate to a valid page, a 404 page will be shown. Is this an intended behaviour ? Can we provide a offline mode page ?

**Comment by MoOx** [comment]
If you have it in page memory you won't have a 404. If you do not have it in memory, this make sense to show a 404 (like any other website/app will). (Btw, same behavior as before this PR).

By offline mode you mean keep page in localstorage? If so we will need to invalidate the cache, that will probably be more work :) But why not.

**Comment by thangngoc89** [comment]
I think the behaviour before this PR is it'll show nothing except an uncatch promise error in console. Showing an 404 error when fetch fail will lead to misunderstood

**Comment by thangngoc89** [comment]
About offline mode,  it should be in another PR imho

**Comment by MoOx** [comment]
If you fetch (via xhr or via a normal link), you will get a 404 anyway. The 404 came from the network error. The only place the 404 are hardcoded is if you call a link that is not know in the collection or if you hit the 404.html

**Comment by thangngoc89** [comment]
Oh right. Got it :)

**Comment by MoOx** [comment]
https://github.com/MoOx/statinamic/blob/82b5ca6b50f0d0247e78befa9520829a8480f4c8/src/fetchJSON/index.js#L20

**Pull request merged**